Nodejs/Product/Nodejs/Workspace/JsFileDebugLaunchTargerProvider.cs

```diff
@@ -0,0 +1,102 @@
+
+        [ImportingConstructor]
+        public JsFileDebugLaunchTargerProvider(SVsServiceProvider serviceProvider, IVsFolderWorkspaceService workspaceService)
+            : base(serviceProvider, workspaceService)
+        {
+        }
+
+        void IVsDebugLaunchTargetProvider.SetupDebugTargetInfo(ref VsDebugTargetInfo vsDebugTargetInfo, DebugLaunchActionContext debugLaunchContext)
+        {
+            var nodeExe = CheckNodeInstalledAndWarn(debugLaunchContext);
+
+            var nodeVersion = Nodejs.GetNodeVersion(nodeExe);
+            if (nodeVersion >= new Version(8, 0) || NodejsProjectLauncher.CheckDebugProtocolOption())
+            {
+                this.SetupDebugTargetInfoForInspectProtocol(ref vsDebugTargetInfo, debugLaunchContext, nodeExe);
+                TelemetryHelper.LogDebuggingStarted("ChromeV2", nodeVersion.ToString(), isProject: false);
+            }
+            else
+            {
+                this.SetupDebugTargetInfoForNodeProtocol(ref vsDebugTargetInfo, debugLaunchContext, nodeExe);
+                TelemetryHelper.LogDebuggingStarted("Node6", nodeVersion.ToString(), isProject: false);
+            }
+        }
+
+        private void SetupDebugTargetInfoForNodeProtocol(ref VsDebugTargetInfo vsDebugTargetInfo, DebugLaunchActionContext debugLaunchContext, string nodeExe)
+        {
+            var target = $"\"{vsDebugTargetInfo.bstrExe}\"";
+            if (debugLaunchContext.LaunchConfiguration.TryGetValue<string>(ScriptArgsKey, out var scriptArgs) && !string.IsNullOrEmpty(scriptArgs))
+            {
+                target += " " + scriptArgs;
+            }
+            vsDebugTargetInfo.bstrArg = target;
+
+            vsDebugTargetInfo.bstrExe = nodeExe;
+            if (debugLaunchContext.LaunchConfiguration.TryGetValue<string>(NodeArgsKey, out var nodeArgs) && !string.IsNullOrWhiteSpace(nodeArgs))
+            {
+                AppendOption(ref vsDebugTargetInfo, AD7Engine.InterpreterOptions, nodeArgs);
+            }
+
+            if (debugLaunchContext.LaunchConfiguration.TryGetValue<string>(DebuggerPortKey, out var debuggerPort) && !string.IsNullOrEmpty(debuggerPort))
+            {
+                AppendOption(ref vsDebugTargetInfo, AD7Engine.DebuggerPort, debuggerPort);
+            }
+            AppendOption(ref vsDebugTargetInfo, AD7Engine.WaitOnNormalExitSetting, "true"); // todo: make this an option
+
+            vsDebugTargetInfo.clsidCustom = AD7Engine.DebugEngineGuid;
+            vsDebugTargetInfo.grfLaunch = (uint)__VSDBGLAUNCHFLAGS.DBGLAUNCH_StopDebuggingOnEnd;
+
+            void AppendOption(ref VsDebugTargetInfo dbgInfo, string option, string value)
+            {
+                if (!string.IsNullOrWhiteSpace(dbgInfo.bstrOptions))
+                {
+                    dbgInfo.bstrOptions += ";";
+                }
+
+                dbgInfo.bstrOptions += option + "=" + HttpUtility.UrlEncode(value);
```

Comment:
>option [](start = 39, length = 6)

We have some reason to believe that the option name doesn't require URL encoding?

Reply:
Hardcoded strings